spec.bs

```diff
@@ -989,6 +997,43 @@ With the above infrastructure in place, we can actually fire and handle the {{Ap
+
+  1. Let |traversable| be |source browsing context|'s [=browsing context/containing navigable=]'s [=navigable/traversable navigable=].
+  1. Let |initiatorOrigin| be |source browsing context|'s [=active document=]'s [=Document/origin=].
+  1. [=parallel queue/enqueue steps|Enqueue the following steps=] to |traversable|'s [=traversable navigable/session history traversal queue=]:
+    1. Let |allSteps| be the result of [=traversable navigable/getting all history steps=] for |traversable|.
+    1. Let |targetStep| be TODO compute using [=traversable navigable/current session history step=], |allSteps|, and |delta|. Will be failure if there is no step |delta| away from the current one.
+    1. If |targetStep| is failure, then return.
+    1. [=Apply the history step=] |targetStep| to |traversable| with true, <var ignore>step</var>, |initiatorOrigin|, and |userInvolvement|.
+</div>
+
+<div algorithm="apply the history step">
+  Modify the <a spec="HTML">apply the history step</a> algorithm as follows. Change <var ignore>checkForUserCancellation</var> to |fireBeforeunloadAndNavigate|. Add the |userInvolvement| parameter. Then, insert the following step after step 12 (which assembles |toTraverse| and other lists) but before step 13 (which checks if unloading is user-cancelled):
+
+  1. If |fireBeforeunloadAndNavigate| is true and the result of <a>firing traversal `navigate` events</a> given |toTraverse|, <var ignore>step</var>, <var ignore>initiatorToCheck</var>, and |userInvolvement| is false, then return.
+</div>
+
+<div algorithm>
+  To <dfn>fire traversal `navigate` events</dfn> for a [=list=] of [=navigables=] |toTraverse|, an integer |step|, an [=origin=] |initiatorOrigin|, and a [=user navigation involvement=] |userInvolvement|:
+
+  1. Let |overallResult| be true.
+  1. Let |totalTasks| be the [=list/size=] of |toTraverse|.
+  1. Let |completedTasks| be 0.
+  1. [=list/For each=] |navigable| of |toTraverse|, [=queue a global task=] on the [=history traversal task source=] given |navigable|'s [=navigable/active document=]'s [=relevant global object=] to run these steps:
+      1. Let |destinationEntry| be the item in the result of [=navigable/getting the session history entries=] for |navigable| that has the greatest [=session history entry/step=] less than or equal to |step|.
+      1. If |destinationEntry|'s [=session history entry/document=] is not equal to |navigable|'s [=navigable/active document=], and |initiatorOrigin| is not [=same origin-domain=] with |navigable|'s [=navigable/active document=]'s [=Document/origin=], then abort these steps.
```

Comment:
Again, this could compare document states rather than documents, but it doesn't really matter.

It would matter in cases where both documents could be null, which would be unintentionally equal.

Reply:
A navigable's active document is never null, right?